membership-attribute-service/app/controllers/SalesforceHookController.scala

```diff
@@ -39,48 +39,65 @@ class SalesforceHookController extends LazyLogging {
-      case -\/(OrgIdMatchingError(orgId)) =>
-        Logger.error(s"Wrong organization Id: $orgId")
-        Future(ApiErrors.unauthorized.copy("Wrong organization Id"))
+  val touchpoint = request.touchpoint
+  val validOrgId = touchpoint.sfOrganisationId
+  val attributeService = touchpoint.attrService
+  implicit val pf = Membership
+
+  def processDeletion(membershipDeletion: MembershipDeletion) = {
+    val userId = membershipDeletion.userId
+    attributeService.delete(userId).map { x =>
+      logger.info(s"Successfully deleted user $userId from ${touchpoint.dynamoTable}.")
+      metrics.put("Delete", 1)
+      Success
+    }.recover { case e: Throwable =>
+      logger.warn(s"Failed to delete user $userId from ${touchpoint.dynamoTable}. Salesforce should retry.", e)
+      Failure
+    }
+  }
+
+  def processUpdate(membershipUpdate: MembershipUpdate) = {
+    val attrs = membershipUpdate.attributes
+    (for {
+      sfId <- OptionT(touchpoint.contactRepo.get(attrs.UserId))
+      membershipSubscription <- OptionT(touchpoint.subService.current[SubscriptionPlan.Member](sfId).map(_.headOption))
+    } yield {
+      val tierFromSalesforceWebhook = attrs.Tier
+      val tierFromZuora = membershipSubscription.plan.charges.benefit.id
+      if (tierFromZuora != tierFromSalesforceWebhook) logger.error(s"Differing tier info for $sfId : sf=$tierFromSalesforceWebhook zuora=$tierFromZuora")
+      attrs.copy(Tier = tierFromZuora)
+    }).run.flatMap { attrsUpdatedWithZuoraOpt =>
+      if (attrsUpdatedWithZuoraOpt.isEmpty) logger.error(s"Couldn't update $attrs with information from Zuora")
+      attributeService.set(attrsUpdatedWithZuoraOpt.getOrElse(attrs)).map { putItemResult =>
+        logger.info(s"Successfully inserted ${attrsUpdatedWithZuoraOpt.getOrElse(attrs)} into ${touchpoint.dynamoTable}.")
+        metrics.put("Update", 1)
+        Success
+      }.recover { case e: Throwable =>
+        logger.warn(s"Failed to insert ${attrsUpdatedWithZuoraOpt.getOrElse(attrs)} into ${touchpoint.dynamoTable}. Salesforce should retry.", e)
+        Failure
+      }
     }
   }
+
+  SFParser.parseOutboundMessage(request.body, validOrgId) match {
+    case -\/(ParsingError(msg)) =>
+      logger.error(s"Could not parse payload. \n$msg")
+      Future(ApiErrors.badRequest(msg))
+    case -\/(OrgIdMatchingError(orgId)) =>
+      logger.error(s"Wrong organization Id: $orgId")
+      Future(ApiErrors.unauthorized.copy("Wrong organization Id"))
+    // On successful parse, we should end up with a Seq of Salesforce objects to update
+    case \/-(outboundMessageChanges @ Seq(_*)) =>
+      logger.info(s"Parsed Salesforce message successfully. Salesforce sent ${outboundMessageChanges.length} objects to update: $outboundMessageChanges")
+      // Take the Seq and apply the appropriate action for each notification item, based on its type
+      val updates = outboundMessageChanges.map( outboundMessage => outboundMessage match {
```

Comment:
you can just miss out the `( outboundMessage => outboundMessage match ` and it automatically assimes you're effectively pattern matching